File: app/src/main/java/com/example/swtrial2/spells/adapterstuff/SpellXmlParser.java
package com.example.swtrial2.spells.adapterstuff;

import android.app.Activity;
import android.content.res.Resources;
import android.content.res.XmlResourceParser;

import org.xmlpull.v1.XmlPullParser;
import org.xmlpull.v1.XmlPullParserException;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;

public class SpellXmlParser {

    public ArrayList<Spell> parse(Activity activity, int xmlResourceId) throws XmlPullParserException, IOException{

        Resources res = activity.getResources();
        XmlResourceParser spelldataXmlParser = res.getXml(xmlResourceId);

        ArrayList<Spell> spelllist = new ArrayList<>();
        ArrayList<String> xmlTagStack = new ArrayList<>();
        Spell currentspell;
        ArrayList<String> spelldata = new ArrayList<>(Arrays.asList("1","2","3","4","5","6","7","8","9","10"));
        spelldataXmlParser.next();
        int eventType= spelldataXmlParser.getEventType();
        while(eventType != XmlPullParser.END_DOCUMENT){
            //Start Tag
            if(eventType == XmlPullParser.START_TAG){
                String tagName = spelldataXmlParser.getName();
                xmlTagStack.add(tagName);
                /*if(tagName.contains("spell")){
                    spelldata.forEach(it-> spelldata.set(spelldata.indexOf(it),"mple"));
                }*/
            }
            //End Tag
            else if (eventType == XmlPullParser.END_TAG){
                String tagName = spelldataXmlParser.getName();
                xmlTagStack.remove(xmlTagStack.size()-1);
                if(tagName.equals("spell")){
                    currentspell = new Spell(spelldata.get(0),spelldata.get(1),spelldata.get(2),spelldata.get(3),Integer.parseInt(spelldata.get(4)),Boolean.parseBoolean(spelldata.get(5)),Boolean.parseBoolean(spelldata.get(6)),Boolean.parseBoolean(spelldata.get(7)),spelldata.get(8),spelldata.get(9));
                    spelllist.add(currentspell);
                }
            }
            //text
            else if (eventType == XmlPullParser.TEXT){
                String currentTag = xmlTagStack.get(xmlTagStack.size()-1);
                String text = spelldataXmlParser.getText();
                switch (currentTag) {
                    case "name":
                        spelldata.set(0, text);
                        break;
                    case "printname":
                        spelldata.set(1, text);
                        break;
                    case "castingTime":
                        spelldata.set(2, text);
                        break;
                    case "side":
                        spelldata.set(3, text);
                        break;
                    case "level":
                        spelldata.set(4, text);
                        break;
                    case "concentration":
                        spelldata.set(5, text);
                        break;
                    case "prerequisite":
                        spelldata.set(6, text);
                        break;
                    case "detailsText":
                        spelldata.set(8, text);
                        break;
                    case "isBig":
                        spelldata.set(7, text);
                        break;
                    case "levelInFull":
                        spelldata.set(9, text);
                        break;

                }
            }
            eventType= spelldataXmlParser.next();
        }

        return spelllist;
    }
}

File: app/src/main/java/com/example/swtrial2/spells/adapterstuff/SpellXmlParser.kt
package com.example.swtrial2.spells.adapterstuff

import android.app.Activity
import android.graphics.Color
import android.text.Spannable
import android.text.SpannableString
import android.text.style.ForegroundColorSpan
import androidx.core.text.buildSpannedString
import org.xmlpull.v1.XmlPullParser
import org.xmlpull.v1.XmlPullParserException
import java.io.IOException
import java.util.*

class SpellXmlParser {
    @Throws(XmlPullParserException::class, IOException::class)
    fun parse(activity: Activity, xmlResourceId: Int): ArrayList<Spell> {
        val res = activity.resources
        val spelldataXmlParser = res.getXml(xmlResourceId)
        val spelllist = ArrayList<Spell>()
        val xmlTagStack = ArrayList<String>()
        var currentspell: Spell
        val spelldata = ArrayList(Arrays.asList("1", "2", "3", "4", "5", "6", "7", "8", "9", "10"))
        spelldataXmlParser.next()
        var eventType = spelldataXmlParser.eventType
        val gold=Color.parseColor("#D4af37")
        while (eventType != XmlPullParser.END_DOCUMENT) {
            //Start Tag
            if (eventType == XmlPullParser.START_TAG) {
                val tagName = spelldataXmlParser.name
                xmlTagStack.add(tagName)
            } else if (eventType == XmlPullParser.END_TAG) {
                val tagName = spelldataXmlParser.name
                xmlTagStack.removeAt(xmlTagStack.size - 1)
                if (tagName == "spell") {
                    currentspell = Spell(spelldata[0],spelldata[1],spelldata[2],spelldata[3],spelldata[4].toInt(),spelldata[5].toBoolean(),spelldata[6].toBoolean(),spelldata[7].toBoolean(),spelldata[8],spelldata[9])
                    val editedspell = currentspell
                    with(currentspell.side){
                        when{
                            this.contains("Dark")->{
                                editedspell.printedname = buildSpannedString {val spanit=SpannableString(currentspell.printedname);spanit.setSpan(ForegroundColorSpan(Color.RED),0,currentspell.printedname.length, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);append(spanit)}
                                editedspell.side= buildSpannedString {val spanit=SpannableString(currentspell.side);spanit.setSpan(ForegroundColorSpan(Color.RED),0,4, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);append(spanit)}
                            }
                            this.contains("Light")->{
                                editedspell.printedname= buildSpannedString {val spanit=SpannableString(currentspell.printedname);spanit.setSpan(ForegroundColorSpan(Color.CYAN),0,currentspell.printedname.length, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);append(spanit)}
                                editedspell.side= buildSpannedString {val spanit=SpannableString(currentspell.side);spanit.setSpan(ForegroundColorSpan(Color.CYAN),0,5, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);append(spanit)}
                            }
                            this.contains("Universal")->{
                                editedspell.printedname= buildSpannedString {val spanit=SpannableString(currentspell.printedname);spanit.setSpan(ForegroundColorSpan(gold),0,currentspell.printedname.length, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);append(spanit)}
                                editedspell.side= buildSpannedString {val spanit=SpannableString(currentspell.side);spanit.setSpan(ForegroundColorSpan(gold),0,this.length, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);append(spanit)}
                            }
                            else->{}
                        }
                    }

                    spelllist.add(editedspell)
                }
            } else if (eventType == XmlPullParser.TEXT) {
                val currentTag = xmlTagStack[xmlTagStack.size - 1]
                val text = spelldataXmlParser.text
                when (currentTag) {
                    "name" -> spelldata[0] = text
                    "printname" -> spelldata[1] = text
                    "castingtime" -> spelldata[2] = text
                    "side" -> spelldata[3] = text
                    "level" -> spelldata[4] = text
                    "concentration" -> spelldata[5] = text
                    "prerequisite" -> spelldata[6] = text
                    "isBig" -> spelldata[7] = text
                    "levelInFull" -> spelldata[8] = text
                    "detailsText" -> spelldata[9] = text
                }
            }
            eventType = spelldataXmlParser.next()
        }
        return spelllist
    }
}